### PR TITLE
Pepper 419 Pediatric Consent Form and SimpleActivityRevisionTask

### PR DESCRIPTION
**Pull request by rfricke-asymmetrik**
## Context

_Say a few words about the bigger picture beyond this PR. How do your changes improve life for our users? Link directly to the jira ticket (just type "DDP-..." and it'll automatically link) to help your teammates quickly get up to speed on the motivation for these changes._

https://broadworkbench.atlassian.net/browse/PEPPER-419
https://broadworkbench.atlassian.net/browse/PEPPER-449

This PR applies changes to the `PARENTAL_CONSENT` and `CONSENT_ASSENT` activities for Osteo. It also adds a new, more generic class for applying these types of changes in bulk for a study. All that is needed is that study builder is run in the context of the study and a file for that study with a list of activity updates similar to the file provided in this PR is sent as an argument to the task. This can be done in the cli with `-- path/to/file` or in java: 
`CustomTask simpleTask = new SimpleActivityRevisionTask();
  simpleTask.init(cfgPath, studyCfg, varsCfg);
  simpleTask.consumeArguments(Array.of("path/to/file"));
  simpleTask.run(handle);`

## Checklist

- [x] I have labeled the type of changes involved using the `C-*` labels.
- [x] I have assessed potential risks and labeled using the `R-*` labels.
- [x] I have considered error handling and alerts, and added `L-*` labels as needed.
- [x] I have considered security and privacy, and added `I-*` labels as needed
- [x] I have analyzed my changes for stability, fault tolerance, graceful degradation, performance bottlenecks and written a brief summary in this PR.
- [ ] If applicable, I have discussed the analytics needs at both a platform and study level with Product and instrumented code accordingly.
- [ ] If applicable, my UI/UX changes have passed muster with Product/Design via an over-the-shoulder review, screenshots, etc.

_If unsure or need help with any of the above items, add the `help wanted` label. For items that starts with `If applicable`, if it is not applicable, check it off and add `n/a` in front._

## FUD Score

_Overall, how are you feeling about these changes?_

- [ ] :relaxed: All good, business as usual!
- [ ] :sweat_smile: There might be some issues here
- [ ] :scream: I'm sweaty and nervous

## How do we demo these changes?

_How does one observe these changes in a deployed system? Note that **user visible** encompasses many personas--not just patients and study staff, but also ops duty, your fellow devs, compliance, etc._

- [x] They are user-visible in dev as a regular user journey and require no additional instructions.
- [ ] Getting dev into a state where this is user-visible requires some tech fiddling. I have documented these steps in the related ticket.
- [ ] Requires other features before it's human visible. I have documented the blocking issues in jira.
- [ ] I have no idea how to demo this. Please help me!

## Testing

- [ ] I have written automated positive tests
- [ ] I have written automated negative tests
- [x] I have written zero automated tests but have poked around locally to verify proper functionality
- [ ] The jira ticket has acceptance criteria and QA has the needed information to test changes

## Release

- [ ] These changes require no special release procedures--just code!
- [ ] Releasing these changes requires special handling and I have documented the special procedures in the release plan document

### CODE REVIEW

File: pepper-apis/studybuilder-cli/studies/osteo/patches/osteo-version-3-changes.conf [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+      "variables-update": [
+        {
+          "name": "v2_osteo_consent_minors_s1_participation_detail_p6",
+          "translations": [{
+              "language": "en",
+              "text": """The project is also partnering with Invitae and Genome Medical in order to sequence and share information on germline DNA (“normal” DNA that is not from your child’s tumor) from saliva samples. After we’ve received your child’s saliva or blood as well as their tumor samples, we’ll reach out to tell you more about this opportunity."""
+            }]
+        },
+        {
+          "name": "v2_osteo_consent_minors_s1_voluntary_content",
```
Comment (aweng98) [inline pepper-apis/studybuilder-cli/studies/osteo/patches/osteo-version-3-changes.conf:18]: Maybe this is an out of scope question. Some variable names begin with `v2-` so it's kind misleading or wrong.

File: pepper-apis/studybuilder-cli/studies/osteo/patches/osteo-version-3-changes.conf [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+            "text": """<ul class="ul-flat-list">
+            <li>Katie Janeway, MD</li>
+            </ul>"""
+          }]
+        },
+        {
+          "name": "osteo_consent_assent_s4_child_intro_detail_p3",
+          "translations": [{
+            "language": "en",
+            "text": """If you choose to share blood samples(s), there are small risks from getting a sample of blood. You may have pain and swelling at the site of the blood draw. <br/> <br/> Your parent(s)/guardian(s) may also be asked if you both want to be interviewed about what it’s like to be part of this research study. You’ll be able to tell your parent(s)/guardian(s) if you do or do not want to give an interview."""
```
Comment (aweng98) [inline pepper-apis/studybuilder-cli/studies/osteo/patches/osteo-version-3-changes.conf]: Just saw this, `samples(s)` is got to be a typo.

File: pepper-apis/studybuilder-cli/src/main/java/org/broadinstitute/ddp/studybuilder/task/SimpleActivityRevisionTask.java [hunk 3]
```diff
@@ -0,0 +1,335 @@
...
+            log.info("terminated: language: {} revId: {} new Text: {} currText: {}", language, currTranslation.getRevisionId(),
+                    newTxt, currentText);
+
+            substitutionIds.add(currTranslation.getId().get());
+            jdbiVarSubst.insert(currTranslation.getLanguageCode(), templateVariable.getTranslation(language).get().getText(),
+                    version.getRevId(), templateVariableId);
+        }
+        int[] ids = jdbiVarSubst.bulkUpdateRevisionIdsBySubIds(substitutionIds, revIds);
+        if (ids.length != revIds.length) {
+            throw new DDPException("returned ids length " + ids.length + "  doesnt match revIds passed length " + revIds.length);
```
Comment (aweng98) [inline pepper-apis/studybuilder-cli/src/main/java/org/broadinstitute/ddp/studybuilder/task/SimpleActivityRevisionTask.java]: recommend `new DDPException(String.format((...);` for code consistency.

File: pepper-apis/studybuilder-cli/src/main/java/org/broadinstitute/ddp/studybuilder/task/SimpleActivityRevisionTask.java [hunk 4]
```diff
@@ -0,0 +1,335 @@
...
+        }
+
+        log.info("Updated Question Template Variables");
+    }
+
+    private void revisionVariableTranslation(TemplateVariable templateVariable,
+                                             RevisionMetadata meta, ActivityVersionDto version) {
+        log.info("revisioning and updating template variable: {}", templateVariable.getName());
+        Long templateVariableId = sqlHelper.findVariableIdByNameAndActivityId(templateVariable.getName(), version.getActivityId());
+        log.info("Tmpl variableId  {} ", templateVariableId);
```
Comment (aweng98) [inline pepper-apis/studybuilder-cli/src/main/java/org/broadinstitute/ddp/studybuilder/task/SimpleActivityRevisionTask.java]: still need this log.info because of line #279?

File: pepper-apis/studybuilder-cli/src/main/java/org/broadinstitute/ddp/studybuilder/task/SimpleActivityRevisionTask.java [hunk 5]
```diff
@@ -0,0 +1,335 @@
...
+        try {
+            configList = activityConfig.getConfigList(BLOCK_UPDATES);
+        } catch (Exception e) {
+            log.info("No Template updates found for activity={}", version.getActivityId());
+            return;
+        }
+
+        for (Config config : configList) {
+            revisionContentBlockTemplate(meta, version, config);
+        }
```
Comment (aweng98) [inline pepper-apis/studybuilder-cli/src/main/java/org/broadinstitute/ddp/studybuilder/task/SimpleActivityRevisionTask.java:170]: Do you want to add another `log.info` at the end of this function to indicate operation has completed? Like in `updateQuestionTemplateVariables()`.

**Review by aweng98 — COMMENTED** [review]
LGTM. Logs are a little inconsistent for me, few are all uppercase. Can you also update the README about this class and its usage with an example?

File: pepper-apis/studybuilder-cli/src/main/java/org/broadinstitute/ddp/studybuilder/task/SimpleActivityRevisionTask.java [hunk 6]
```diff
@@ -0,0 +1,335 @@
...
+import org.broadinstitute.ddp.util.GsonUtil;
+import org.jdbi.v3.core.Handle;
+import org.jdbi.v3.sqlobject.SqlObject;
+import org.jdbi.v3.sqlobject.config.RegisterConstructorMapper;
+import org.jdbi.v3.sqlobject.customizer.Bind;
+import org.jdbi.v3.sqlobject.statement.SqlQuery;
+
+@Slf4j
+public class SimpleActivityRevisionTask implements CustomTask {
+    private static final String VARIABLES_UPD = "variables-update";
```
Comment (aweng98) [inline pepper-apis/studybuilder-cli/src/main/java/org/broadinstitute/ddp/studybuilder/task/SimpleActivityRevisionTask.java]: `VARIABLES_UPD` vs `STUDY_UPDATES`: which is better for keeping? abbriviation UPD or UPDATES

File: pepper-apis/studybuilder-cli/src/main/java/org/broadinstitute/ddp/studybuilder/task/SimpleActivityRevisionTask.java [hunk 7]
```diff
@@ -0,0 +1,335 @@
...
+        log.info(positional[0]);
+        this.dataFile = positional[0];
+
+        File file = cfgPath.getParent().resolve(this.dataFile).toFile();
+        if (!file.exists()) {
+            throw new DDPException("Data file is missing: " + file);
+        }
+        dataCfg = ConfigFactory.parseFile(file).resolveWith(varsCfg);
+
+        cfg = studyCfg;
```
Comment (aweng98) [inline pepper-apis/studybuilder-cli/src/main/java/org/broadinstitute/ddp/studybuilder/task/SimpleActivityRevisionTask.java]: Is `cfg` used at all? it's same as `this.studyCfg`

File: pepper-apis/studybuilder-cli/src/main/java/org/broadinstitute/ddp/studybuilder/task/SimpleActivityRevisionTask.java [hunk 8]
```diff
@@ -0,0 +1,335 @@
...
+    private void updateTemplates(RevisionMetadata meta, ActivityVersionDto version, Config activityConfig) {
+        log.info("UPDATE Templates/Content Blocks");
+        List<? extends Config> configList = null;
+
+        try {
+            configList = activityConfig.getConfigList(BLOCK_UPDATES);
+        } catch (Exception e) {
+            log.info("No Template updates found for activity={}", version.getActivityId());
+            return;
+        }
```
Comment (aweng98) [inline pepper-apis/studybuilder-cli/src/main/java/org/broadinstitute/ddp/studybuilder/task/SimpleActivityRevisionTask.java:166]: What do you think of a change?
If I initialize `List<? extends Config> configList` early in class (for example, in run() function), and if  `getConfigList` throws an exception, execution would stops early before calling and making any updates. I'm afraid if it throws an exeception  here, something are updated while something else are not.

File: pepper-apis/studybuilder-cli/src/main/java/org/broadinstitute/ddp/studybuilder/task/SimpleActivityRevisionTask.java [hunk 9]
```diff
@@ -0,0 +1,335 @@
...
+        this.studyCfg = studyCfg;
+    }
+
+    @Override
+    public void run(Handle handle) {
+        this.adminUser = handle.attach(UserDao.class).findUserByGuid(cfg.getString("adminUser.guid")).get();
+
+        this.studyDto = handle.attach(JdbiUmbrellaStudy.class).findByStudyGuid(dataCfg.getString("study.guid"));
+
+        List<? extends Config> activityUpdateConfigs = dataCfg.getConfigList(STUDY_UPDATES);
```
Comment (aweng98) [inline pepper-apis/studybuilder-cli/src/main/java/org/broadinstitute/ddp/studybuilder/task/SimpleActivityRevisionTask.java]: `dataCfg.getConfigList` does not handle exception. And `List<? extends Config> configList` could be initialized here.

File: pepper-apis/studybuilder-cli/src/main/java/org/broadinstitute/ddp/studybuilder/task/SimpleActivityRevisionTask.java (hunk 8 above)
Comment (rfricke-asymmetrik) [reply pepper-apis/studybuilder-cli/src/main/java/org/broadinstitute/ddp/studybuilder/task/SimpleActivityRevisionTask.java:166]: In this case, it is getting a list of configs off of a nested object in the json specific to the activity. `updateActivities` is called with the list that is found from `activity-updates` which then loops through those and passes that config down into each of the three separate methods where they then grab the list of changes off of that config object.

Btw, when a task is run under the StudyBuilderCli, the whole task run is wrapped in a transaction so, with a failure, changes are rolled back. All pass or none pass.

Comment (otchet-broad) [reply pepper-apis/studybuilder-cli/src/main/java/org/broadinstitute/ddp/studybuilder/task/SimpleActivityRevisionTask.java:166]: Building on what Rich said, this is intending to be a generic update processor and although this exemplar includes a Templates/Content Block,  it shouldn't have to for the other types of updates to run.  Same holds true for each of the other types of updates this supports.

Comment (aweng98) [reply pepper-apis/studybuilder-cli/src/main/java/org/broadinstitute/ddp/studybuilder/task/SimpleActivityRevisionTask.java:166]: I see. I like this a lot. Thank you both for making this happen.

File: pepper-apis/studybuilder-cli/studies/osteo/patches/osteo-version-3-changes.conf [hunk 10]
```diff
@@ -0,0 +1,273 @@
...
+  "study": {
+    "guid": "CMI-OSTEO"
+  },
+  "activity-updates": [
+    {
+      "versionTag": "v3",
+      "activityCode": "PARENTAL_CONSENT",
+      "variable-updates": [
+        {
+          "name": "v2_osteo_consent_minors_s1_participation_detail_p6",
```
Comment (ssettipalli) [inline pepper-apis/studybuilder-cli/studies/osteo/patches/osteo-version-3-changes.conf:11]: Do we see any additional advantage by renaming the variable and adding v2_ ?

**Comment by ssettipalli** [comment]
Please add this patch to osteo pe-cgs rootpatch: Osteo2pecgsUpdates.java

File: pepper-apis/studybuilder-cli/studies/osteo/patches/osteo-version-3-changes.conf (hunk 10 above)
Comment (rfricke-asymmetrik) [reply pepper-apis/studybuilder-cli/studies/osteo/patches/osteo-version-3-changes.conf:11]: This is what the variable is named currently. Are you saying we should rename the variable to v3_?

Comment (ssettipalli) [reply pepper-apis/studybuilder-cli/studies/osteo/patches/osteo-version-3-changes.conf:11]: If thats current name, I am good. No renaming needed . Thanks

**Review by aweng98 — APPROVED** [review]
LGTM! It turned into a much bigger PR than planned. :)

File: pepper-apis/studybuilder-cli/src/main/java/org/broadinstitute/ddp/studybuilder/task/rootpatches/Osteo2pecgsUpdates.java [hunk 11]
```diff
@@ -30,7 +33,17 @@ public void init(Path cfgPath, Config studyCfg, Config varsCfg) {
...
         taskList.add(new Osteo2GermlinePedConsentVersion3());
 
+        SimpleActivityRevisionTask osteoPediatricConsentAndAssentVersion3 = new SimpleActivityRevisionTask();
+        taskList.add(osteoPediatricConsentAndAssentVersion3);
+
         taskList.forEach(task -> task.init(cfgPath, studyCfg, varsCfg));
+
+        try {
+            osteoPediatricConsentAndAssentVersion3.consumeArguments(
+                    new String[]{"patches/osteo-version-3-changes.conf"});
```
Comment (ssettipalli) [inline pepper-apis/studybuilder-cli/src/main/java/org/broadinstitute/ddp/studybuilder/task/rootpatches/Osteo2pecgsUpdates.java:43]: Just a comment : From my understanding :
Adding incremental config changes to same config  (like patches/osteo-version-3-changes.conf)  file will not work in Dev env but will work in rest of envs because we apply All config changes at once. In Dev though as we complete each ticket (new config change), we apply in DEV.  I mean previous config changes might have been already applied in Dev.

**Review by ssettipalli — APPROVED** [review]
looks good to me

File: pepper-apis/studybuilder-cli/src/main/java/org/broadinstitute/ddp/studybuilder/task/rootpatches/Osteo2pecgsUpdates.java (hunk 11 above)
Comment (ssettipalli) [reply pepper-apis/studybuilder-cli/src/main/java/org/broadinstitute/ddp/studybuilder/task/rootpatches/Osteo2pecgsUpdates.java:43]: probably making a new patch config file and adding new entry might be workaround for Dev scenario.

Comment (otchet-broad) [reply pepper-apis/studybuilder-cli/src/main/java/org/broadinstitute/ddp/studybuilder/task/rootpatches/Osteo2pecgsUpdates.java:43]: That's right.  We could decide in the future to change the behavior and incrementally apply the changes.
